network/server.go

```diff
@@ -376,15 +407,38 @@ func (s *Server) hasPeer(peerID peer.ID) bool {
 	return ok
 }
 
-func (s *Server) numOpenSlots() int64 {
-	n := int64(s.config.MaxPeers) - (s.numPeers() + s.identity.numPending())
+func (s *Server) availableOutboundConns() int64 {
+	n := s.maxOutboundConns() - s.outboundConns()
 	if n < 0 {
 		n = 0
 	}
 
 	return n
 }
 
+func (s *Server) inboundConns() int64 {
+	count := atomic.LoadInt64(&s.inboundConnCount)
+	if count < 0 {
+		count = 0
+	}
+
+	return count + s.identity.pendingInboundConns()
+}
+
+func (s *Server) outboundConns() int64 {
+	activeOutboundConns := s.numPeers() - atomic.LoadInt64(&s.inboundConnCount)
+
+	return activeOutboundConns + s.identity.pendingOutboundConns()
+}
```

Comment:
Honestly, I don't understand this. `s.numPeers()` returns the length of the `peers` field of type `map[peer.ID]*Peer` in the `Server` struct. Peers are identified by `ID` field, so how does this work `activeOutboundConns := s.numPeers() - atomic.LoadInt64(&s.inboundConnCount)` if one peer can have both inbound/outbound connections? `s.numPeers()` isn't the same as the number of total connections in my understanding.